src/controller/InteractionPeer.cpp

```diff
@@ -0,0 +1,121 @@
+                                                     Callback::Cancelable * onFailureCallback, app::TLVDataFilter aTlvDataFilter)
+{
+
+    uint8_t seqNum = GetNextSequenceNumber();
+    aPath.mNodeId  = mDelegate->GetDeviceId();
+
+    ReturnErrorOnFailure(mDelegate->PreparePeer());
+
+    if (onSuccessCallback != nullptr || onFailureCallback != nullptr)
+    {
+        AddResponseHandler(seqNum, onSuccessCallback, onFailureCallback, aTlvDataFilter);
+    }
+    // The application context is used to identify different requests from client applicaiton the type of it is intptr_t, here we
+    // use the seqNum.
+    CHIP_ERROR err = chip::app::InteractionModelEngine::GetInstance()->SendReadRequest(
+        mDelegate->GetDeviceId(), 0, mDelegate->GetSesssionHandle(), nullptr /*event path params list*/, 0, &aPath, 1,
+        0 /* event number */, seqNum /* application context */);
+    if (err != CHIP_NO_ERROR)
+    {
+        CancelResponseHandler(seqNum);
+    }
+    return err;
+}
+
+CHIP_ERROR InteractionPeer::SendWriteAttributeRequest(app::WriteClientHandle aHandle, Callback::Cancelable * onSuccessCallback,
+                                                      Callback::Cancelable * onFailureCallback)
+{
+    uint8_t seqNum = GetNextSequenceNumber();
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    aHandle->SetAppIdentifier(seqNum);
+    ReturnErrorOnFailure(mDelegate->PreparePeer());
+
+    if (onSuccessCallback != nullptr || onFailureCallback != nullptr)
+    {
+        AddResponseHandler(seqNum, onSuccessCallback, onFailureCallback);
+    }
+    if ((err = aHandle.SendWriteRequest(mDelegate->GetDeviceId(), 0, mDelegate->GetSesssionHandle())) != CHIP_NO_ERROR)
+    {
+        CancelResponseHandler(seqNum);
+    }
+    return err;
+}
+
+void InteractionPeer::AddResponseHandler(uint8_t seqNum, Callback::Cancelable * onSuccessCallback,
+                                         Callback::Cancelable * onFailureCallback, app::TLVDataFilter tlvDataFilter)
+{
+    mCallbacksMgr.AddResponseCallback(mDelegate->GetDeviceId(), seqNum, onSuccessCallback, onFailureCallback, tlvDataFilter);
+}
+
+void InteractionPeer::CancelResponseHandler(uint8_t seqNum)
+{
+    mCallbacksMgr.CancelResponseCallback(mDelegate->GetDeviceId(), seqNum);
+}
+
+void InteractionPeer::AddIMResponseHandler(app::CommandSender * commandObj, Callback::Cancelable * onSuccessCallback,
+                                           Callback::Cancelable * onFailureCallback)
+{
+    // We are using the pointer to command sender object as the identifier of command transactions. This makes sense as long as
+    // there are only one active command transaction on one command sender object. This is a bit tricky, we try to assume that
+    // chip::NodeId is uint64_t so the pointer can be used as a NodeId for CallbackMgr.
+    static_assert(std::is_same<chip::NodeId, uint64_t>::value, "chip::NodeId is not uint64_t");
```

Comment:
@pan-apple this is a multi-admin issue I think, right here. The fact that NodeID uniqueness is globally expected across different fabrics for callback tracking properly.

Reply:
This function is moved from the `CHIPDevice` class. I believe it's okay to have the comment here.